exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/entities/ViaTest.kt

```diff
@@ -288,74 +286,114 @@ class ViaTests : DatabaseTestsBase() {
-
     class Project(id: EntityID<Int>) : IntEntity(id) {
         companion object : IntEntityClass<Project>(Projects)
 
         var name by Projects.name
-        val tasks by Task via ProjectTasks
+        var tasks by TaskWithData via ProjectTasks
     }
 
-    object ProjectTasks : CompositeIdTable("project_tasks") {
+    object ProjectTasks : Table("project_tasks") {
         val project = reference("project", Projects, onDelete = ReferenceOption.CASCADE)
         val task = reference("task", Tasks, onDelete = ReferenceOption.CASCADE)
         val approved = bool("approved")
+        val sprint = integer("sprint")
 
         override val primaryKey = PrimaryKey(project, task)
-
-        init {
-            addIdColumn(project)
-            addIdColumn(task)
-        }
-    }
-
-    class ProjectTask(id: EntityID<CompositeID>) : CompositeEntity(id) {
-        companion object : CompositeEntityClass<ProjectTask>(ProjectTasks)
-
-        var approved by ProjectTasks.approved
     }
 
-    object Tasks : IntIdTable("tasks") {
+    // IdTable without auto-increment is used so manual ids can be inserted without excluding SQL Server
+    object Tasks : IdTable<Int>("tasks") {
+        override val id = integer("id").entityId()
         val title = varchar("title", 64)
+        override val primaryKey = PrimaryKey(id)
     }
-
     class Task(id: EntityID<Int>) : IntEntity(id) {
         companion object : IntEntityClass<Task>(Tasks)
 
         var title by Tasks.title
-        val approved by ProjectTasks.approved
+        var projects by ProjectWithData via ProjectTasks
+    }
+
+    class ProjectWithData(
+        val project: Project,
+        val approved: Boolean,
+        val sprint: Int
+    ) : InnerTableLinkEntity<Int>(project) {
+        companion object : InnerTableLinkEntityClass<Int, ProjectWithData>(Projects) {
+            override fun createInstance(entityId: EntityID<Int>, row: ResultRow?): ProjectWithData {
+                return row?.let {
+                    ProjectWithData(Project.wrapRow(it), it[ProjectTasks.approved], it[ProjectTasks.sprint])
+                } ?: ProjectWithData(Project(entityId), false, 0)
+            }
+        }
+
+        override fun getInnerTableLinkValue(column: Column<*>): Any = when (column) {
+            ProjectTasks.approved -> approved
+            ProjectTasks.sprint -> sprint
+            else -> error("Column does not exist in intermediate table")
+        }
+    }
+
+    class TaskWithData(
+        val task: Task,
+        val approved: Boolean,
+        val sprint: Int
+    ) : InnerTableLinkEntity<Int>(task) {
+        companion object : InnerTableLinkEntityClass<Int, TaskWithData>(Tasks) {
+            override fun createInstance(entityId: EntityID<Int>, row: ResultRow?): TaskWithData {
+                return row?.let {
+                    TaskWithData(Task.wrapRow(it), it[ProjectTasks.approved], it[ProjectTasks.sprint])
+                } ?: TaskWithData(Task(entityId), false, 0)
```

Comment:
Must user handle null rows here?

As I can see from tests the instances of `TaskWithData` are created inside the application logic. When 2 entities are connected user can create link with extra data between them.

Probably I just missed something, but it's not clear for me, when I should expect the call of `createdInstance` with null row.

I replaced `?: TaskWithData(Task(entityId), false, 0)` with `?: error("")` and tests continued to be green.

Reply:
I agree that I'm not a fan of the null handling.
`createInstance()` is an existing open method of `EntityClass` that I chose to force an override for, since it's the important function that is called every time we try to wrap a row for a link entity that doesn't already exist in the cache.

I could create a new method/overload that doesn't have a nullable row, which just calls `createInstance()` under the hood and we handle null ourselves.